### PR TITLE
RATEST-80: Failing Test: Visit Note Test

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-80. cc @dkayiwa  @ibacher @kdaud

### CODE REVIEW

**Comment by sherrif10** [comment]
For this ticket am still looking at the clearer way of handling some automations especially with edit, delete and view, we have two ways of handling these three features.

**Comment by sherrif10** [comment]
Working on resolving the build faiures

**Comment by sherrif10** [comment]
Hello @k-joseph  Requesting for review

**Comment by k-joseph** [comment]
> For this ticket am still looking at the clearer way of handling some automations especially with edit, delete and view, we have two ways of handling these three features.

Any progress on this?

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:28]: There are trailing spaces before private too, you can get rid of those by pressing the cursor on the top line, press enter and immediately start typing vs using the cursor at the start of the line and using spacing to fix indention spacing

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:36]: any reason for creating this object?

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:74]: this should be dropped, it's unused and doesn't sound a right action to qualify for a method

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:33]: Does visit Note have both edit and view functionalities? you could be instead landing on ones for the listed encounters

**Comment by sherrif10** [comment]
> > For this ticket am still looking at the clearer way of handling some automations especially with edit, delete and view, we have two ways of handling these three features.
> 
> Any progress on this?

Sure i finally got way of handling these automations ,Since the key factors to use are edit, delete and view, one of the ways worked

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:18]: Do you want this to drop both primary and secondary at the same time, this current would only pick the secondary one? 
Rather use something more reliable such as;
`i[data-ng-click='removeDiagnosis(d)']`
or
`.icon-remove.delete-item`

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:33]: So visit Note comprises of view button ,edit and delete however for view button, it have other similar buttons which we are not going to use by default but in actual sense those buttons can also delete and edit as well

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:74]: True we can get rid of it since it does the same functionality with **EDIT_VISIT_NOTE**

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java]: Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java]: Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java]: No Reason thanks

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:29]: Resolved thanks

**Comment by sherrif10** [comment]
Hello @k-joseph  is this worth it

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (k-joseph) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:33]: could you take a screenshot and share it here?

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:33]: visitNotePage
So here is real visit Note  , however all the functionalities of visit Note and visits are in visitsPage,

**Comment by sherrif10** [comment]
cc @ibacher  could you also have some time please thanks, This LGTM

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:55]: Why were these asserts removed?

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:56]: Should we maybe break out editing a note as a separate test?

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:62]: Shouldn't we have some more assertions to ensure the note is actually updated?

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (ibacher) [inline ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:76]: Should this be done here or in the teardown? If here, should we split this into another test?

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (ibacher) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:33]: I think most of these CSS selectors are one class too deep, e.g. `#confirm-delete-dialog > div.dialog-content > div > button.confirm` should be sufficient. Ditto for `#encountersList > li > span > i.editEncounter.edit-action`. For edit and view, we might want to add the `:first-of-type` pseudo-selector.

Comment (ibacher) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:33]: Alright, this is a little insane, but I think a better way to get the view-action or edit-action is this:

```java
By.xpath("//ul[@class=\"encounter-details\"]/child::*//i[contains(@class, \"icon-vitals\")][1]/following::*//i[contains(@class, \"view-action\")][1]");
By.xpath("//ul[@class=\"encounter-details\"]/child::*//i[contains(@class, \"icon-vitals\")][1]/following::*//i[contains(@class, \"edit-action\")][1]")
```

This attempts to select the first `.view-action` or `.edit-action` item that comes after the first encounter with a vitals icon. That way, this should work even if there are additional encounter types.

**Review by ibacher — COMMENTED** [review]
@sherrif10 Some comments on this

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:56]: We are trying to make them independent, breaking the editing note would mean some tests will be dependent on each other

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:55]: These actually are breaking the tests,Looking forward to rectify it

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:55]: Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:82]: Thanks @ibacher  , if we split them we it would look somewhat not straight especially maintaining them , so we might need to make them in one single test for easy workflow

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:68]: Resolved thanks

**Comment by sherrif10** [comment]
Hello @ibacher  i tried this use case out but its breaking in the cssSelectors , but i think the above cssSelector elements are worth using

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:33]: I tried rectifying this following the same suggestions but combining them breaks it

**Comment by sherrif10** [comment]
Hello @ibacher  Looking forward to hear from you

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:46]: This is a minor concern though might be worth looking into! I would emit any redundant line after declaring a method. In short I would remove code line  `45`and `46 `for quality assurance.

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:17]: Since `note` is only used within `EditVisitNotePage` class then making it private static final variable is  a better deal.

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:126]: Remove one of the redundant line `125` or `126`

**Comment by kdaud** [comment]
Great work @sherrif10 ! 
Check my recent reviews.

**Comment by sherrif10** [comment]
Resolved your reviews @kdaud  cc @dkayiwa  @ibacher

**Comment by dkayiwa** [comment]
@sherrif10 before pinging us, can you first ask @kdaud to confirm that all his review comments are attended to?

**Comment by sherrif10** [comment]
cc @kdaud

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:39]: Any reason for declaring `patientVisitDashboardPage` object within the `setUp()` function?

File: ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/VisitNoteTest.java:46]: Remove the redundant code Line `LN-46`

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:15]: Any reason for removing the existing empty line after `EditVisitNotePage` class declaration?

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:19]: Should we not add an empty line to separate declared variables and methods  to ensure quality within our code base?

**Comment by kdaud** [comment]
Well-done @sherrif10 ! 

Catch one or two of my reviews above!

**Comment by sherrif10** [comment]
cc @kdaud

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:15]: To have a clean code description

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:20]: Added a new line , though to me as fur as i have learnt openmrs code especially with in openmrs core, this is no where implemented , though am seeing its getting growing with in current qa work, so i dont know whether we  can continue to implement it in any other code with in openmrs cc @dkayiwa  @ibacher

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:20]: I object neither!
However, I want to believe that embracing software standards in our code base is one our role as QA Team to ensure our code base aligns with international software standards. I do not mean that our current code-base does not align with the standards, but there might be some oversights within our code base that we need to work on within our community and to me this sounds to make a difference why QA Team was embraced in the community.

@sherrif10 in reference to my suggestion [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/450/files#r679962005), I gathered some artifacts on this opinion [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/476#issuecomment-870471982)

**Comment by sherrif10** [comment]
cc @kdaud , do you have more reveiws such that we can hear from @dkayiwa   @ibacher  and other

**Comment by dkayiwa** [comment]
@sherrif10 i see you asking @kdaud for reviews. But have you reviewed his?

**Comment by sherrif10** [comment]
Am working on build failures brought about merge conflicts

**Comment by dkayiwa** [comment]
@sherrif10 reviewing each other's work should take higher priority because they are blocked while waiting on your reviews. Imagine if you take a number of days or a week resolving build failures, do they have to wait for that long?

**Comment by sherrif10** [comment]
Ohh sorry will try to pull up on different reviews , though i rarely see folks ping, however will try as much as possible to review them. Thanks @dkayiwa

**Comment by dkayiwa** [comment]
@sherrif10 here is a review which has been waiting for you for 4 days: https://github.com/openmrs/openmrs-distro-referenceapplication/pull/510#issuecomment-889149677

**Comment by sherrif10** [comment]
I honestly didn't see that in the email, otherwise thanks for the reminder, Will keep an eye

**Comment by sherrif10** [comment]
cc @kdaud  @dkayiwa  @ibacher  Looking foward for more  reviews

**Comment by dkayiwa** [comment]
@sherrif10 before pinging us, have you ensured that @kdaud has reviewed and approved?

**Comment by sherrif10** [comment]
cc @kdaud

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:18]: Should `note` be not a final  variable ?

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: Resolved thanks

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: Hello @kdaud  Any more reviews and recommendations

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: @sherrif10 what do you say about this artifacts [here](https://stackoverflow.com/questions/7259687/java-naming-convention-for-static-final-variables) about naming convention of static final variables ?

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (sherrif10) [inline ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:18]: cc @kdaud  did you see this commit

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:18]: The resource I shared [above](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/450#discussion_r681447476), my point was on `note` vs `NOTE`

**Comment by kdaud** [comment]
`One minor comment:` I see most of the reviews on this commit marked as resolved by the PR author. However, IMO I think if a review is to be marked as resolved any way, this should be done by the reviewer **NOT** by the PR author.
 @dkayiwa any recommendation on this ?

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:18]: Adding some thing minor: I would start my string value with Uppercase letter, something like "`This is edited note`."

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java:18]: > my point was on `note` vs `NOTE`

Please there is a difference between note and Note, i used note to differentiate it from NOTE which is line 17.

**Comment by sherrif10** [comment]
cc @dkayiwa  @ibacher

File: ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java
Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: > i used note to differentiate it from NOTE which is line 17.

In java programming variable naming convention a `static final variable` should be in `UPPERCASE`. See the reference again [here](https://stackoverflow.com/questions/7259687/java-naming-convention-for-static-final-variables)

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: Thats  a simply a string whether its upper case or lower case its fine, as long as you use it how you have called it , please check a number of other areas where such kind of work is being used. Using  **note** to have a clear readable difference with NOTE .

Comment (dkayiwa) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: @sherrif10 did you get a chance to look at this? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NamingConventions

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: Yeah , Let me commit it

Comment (dkayiwa) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: `Class and instance variables should follow Java specs. Any constants should be written in uppercase letters.`

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: Resolved thanks @dkayiwa

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/EditVisitNotePage.java]: Resolved Thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (kdaud) [inline ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:124]: `LN-124` should we not remove this empty line? I can see already another empty line existing to cater for contributors git collaboration. If `LN-124` remains we will have two last empty lines which may not be a good thing!

**Comment by kdaud** [comment]
Great effort @sherrif10 
I have added one last minor comment! 
You may catch it and we see this test run in our master branch

File: ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:124]: > I can see already another empty line existing`(behind the seen)` t that caters for contributors git collaboration.
Which line is that, this line 124 is the last line, btw what do you mean by behind scenes ,Does git provides this behind scenes scenarios, there is no line 125. it only stops on line 124

Comment (kdaud) [reply ui-tests/src/test/java/org/openmrs/reference/page/VisitNotePage.java:124]: > btw what do you mean by behind scenes ,Does git provides this behind scenes scenarios

I often use the git functionality of Edit file to observe the empty line!

Screenshot from 2021-08-04 10-49-20

**Comment by sherrif10** [comment]
Thanks @kdaud

**Review by kdaud — APPROVED** [review]
Thanks @sherrif10  for the good work!
I would merge this commit into master!
cc: @dkayiwa @ibacher

**Comment by kdaud** [comment]
This commit could be a ripe candidate for merging! 
cc: @dkayiwa @ibacher

**Comment by sherrif10** [comment]
cc @dkayiwa @ibacher  ,Waiting on your feedback on this

**Review by ibacher — APPROVED** [review]

**Comment by ibacher** [comment]
Thanks @sherrif10!

**Comment by sherrif10** [comment]
Thanks @ibacher

**Pull request merged**
